File: samsara/hsr_fandom.py
import logging
import shutil
from pathlib import Path

import requests as requests

from samsara.fandom import QueryResponse, query_all


def get_event_wishes() -> QueryResponse:
    logging.info("gathering all event wishes")
    return query_all(
        {
            "action": "query",
            "generator": "categorymembers",
            "gcmtitle": "Category:Event_Warps",
            "prop": "categories",
            "cllimit": "max",
            "gcmlimit": "max",
            "format": "json",
        },
        api_url="https://honkai-star-rail.fandom.com/api.php",
    )


def get_5_star_characters() -> QueryResponse:
    logging.info("gathering all 5 star characters")
    return query_all(
        {
            "action": "query",
            "generator": "categorymembers",
            "gcmtitle": "Category:5-Star_Characters",
            "gcmlimit": "max",
            "format": "json",
        },
        api_url="https://honkai-star-rail.fandom.com/api.php",
    )


def get_4_star_characters() -> QueryResponse:
    logging.info("gathering all 4 star characters")
    return query_all(
        {
            "action": "query",
            "generator": "categorymembers",
            "gcmtitle": "Category:4-Star_Characters",
            "gcmlimit": "max",
            "format": "json",
        },
        api_url="https://honkai-star-rail.fandom.com/api.php",
    )


def get_5_star_weapons() -> QueryResponse:
    logging.info("gathering all 5 star weapons")
    return query_all(
        {
            "action": "query",
            "generator": "categorymembers",
            "gcmtitle": "Category:5-Star_Light_Cones",
            "gcmlimit": "max",
            "format": "json",
        },
        api_url="https://honkai-star-rail.fandom.com/api.php",
    )


def get_4_star_weapons() -> QueryResponse:
    logging.info("gathering all 4 star weapons")
    return query_all(
        {
            "action": "query",
            "generator": "categorymembers",
            "gcmtitle": "Category:4-Star_Light_Cones",
            "gcmlimit": "max",
            "format": "json",
        },
        api_url="https://honkai-star-rail.fandom.com/api.php",
    )


def download_character_image(output_path: str | Path, character_name: str, size: int):
    logging.info(f"downloading {character_name} icon to {output_path}")
    # https://honkai-star-rail.fandom.com/index.php?title=Special:Redirect/file/Character%20Hook%20Icon.png
    r = requests.get(
        f"https://honkai-star-rail.fandom.com/index.php?title=Special:Redirect/file/Character {character_name} Icon.png&width={size}&height={size}",
        stream=True,
    )

    if r.status_code != 200:
        logging.warning(
            f"Received status {r.status_code} trying to download character image {character_name}"
        )
    else:
        with open(output_path, "wb") as f:
            r.raw.decode_content = True
            shutil.copyfileobj(r.raw, f)


def download_weapon_image(output_path: str | Path, weapon_name: str, size: int):
    logging.info(f"downloading {weapon_name} icon to {output_path}")

    r = requests.get(
        f"https://honkai-star-rail.fandom.com/index.php?title=Special:Redirect/file/Light Cone {weapon_name}.png&width={size}&height={size}",
        stream=True,
    )

    if r.status_code != 200:
        logging.warning(
            f"Received status {r.status_code} trying to download weapon image {weapon_name}"
        )
    else:
        with open(output_path, "wb") as f:
            r.raw.decode_content = True
            shutil.copyfileobj(r.raw, f)


def get_page_content(page_id: int) -> QueryResponse:
    logging.info(f"fetching page content for {page_id}")
    return query_all(
        {
            "action": "query",
            "prop": "revisions",
            "pageids": str(page_id),
            "rvprop": "content",
            "rvslots": "main",
            "format": "json",
            "formatversion": "2",
        },
        api_url="https://honkai-star-rail.fandom.com/api.php",
    )

File: src/hsr-fandom.ts
import axios from 'axios';
import * as fs from 'fs';
import * as path from 'path';
import { QueryResponse, PageContent } from './types';
import { queryAll } from './fandom';

const HSR_API_URL = 'https://honkai-star-rail.fandom.com/api.php';

export async function getEventWishes(): Promise<QueryResponse> {
  console.log('Gathering all HSR event wishes');
  return queryAll({
    action: 'query',
    generator: 'categorymembers',
    gcmtitle: 'Category:Event_Warps',
    prop: 'categories',
    cllimit: 'max',
    gcmlimit: 'max',
    format: 'json'
  }, HSR_API_URL);
}

export async function get5StarCharacters(): Promise<QueryResponse> {
  console.log('Gathering all HSR 5 star characters');
  return queryAll({
    action: 'query',
    generator: 'categorymembers',
    gcmtitle: 'Category:5-Star_Characters',
    gcmlimit: 'max',
    format: 'json'
  }, HSR_API_URL);
}

export async function get4StarCharacters(): Promise<QueryResponse> {
  console.log('Gathering all HSR 4 star characters');
  return queryAll({
    action: 'query',
    generator: 'categorymembers',
    gcmtitle: 'Category:4-Star_Characters',
    gcmlimit: 'max',
    format: 'json'
  }, HSR_API_URL);
}

export async function get5StarWeapons(): Promise<QueryResponse> {
  console.log('Gathering all HSR 5 star light cones');
  return queryAll({
    action: 'query',
    generator: 'categorymembers',
    gcmtitle: 'Category:5-Star_Light_Cones',
    gcmlimit: 'max',
    format: 'json'
  }, HSR_API_URL);
}

export async function get4StarWeapons(): Promise<QueryResponse> {
  console.log('Gathering all HSR 4 star light cones');
  return queryAll({
    action: 'query',
    generator: 'categorymembers',
    gcmtitle: 'Category:4-Star_Light_Cones',
    gcmlimit: 'max',
    format: 'json'
  }, HSR_API_URL);
}

export async function downloadCharacterImage(outputPath: string, characterName: string, size: number = 80): Promise<void> {
  const filename = generateFilename(characterName);
  const url = `https://static.wikia.nocookie.net/houkai-star-rail/images/thumb/c/c7/${filename}_Icon.png/${size}px-${filename}_Icon.png`;
  
  await downloadImage(url, outputPath);
}

export async function downloadWeaponImage(outputPath: string, weaponName: string, size: number = 80): Promise<void> {
  const filename = generateFilename(weaponName);
  const url = `https://static.wikia.nocookie.net/houkai-star-rail/images/thumb/c/c7/${filename}_Icon.png/${size}px-${filename}_Icon.png`;
  
  await downloadImage(url, outputPath);
}

async function downloadImage(url: string, outputPath: string): Promise<void> {
  try {
    const response = await axios.get(url, { responseType: 'stream' });
    
    // Ensure directory exists
    const dir = path.dirname(outputPath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    
    const writer = fs.createWriteStream(outputPath);
    response.data.pipe(writer);
    
    return new Promise((resolve, reject) => {
      writer.on('finish', resolve);
      writer.on('error', reject);
    });
  } catch (error) {
    console.warn(`Failed to download HSR image from ${url}: ${error}`);
  }
}

function generateFilename(name: string): string {
  return name
    .replace(/[^a-zA-Z0-9\s]/g, '')
    .replace(/\s+/g, '_')
    .toLowerCase();
}
